### PR TITLE
Feature/28 dogu configuration

### LINKED ISSUE #28: Resources like cpu, memory and volume size aren't configurable

### PR DESCRIPTION
**Pull request by Subaro**
Resolves #28

### CODE REVIEW

File: controllers/dogu_manager_external_test.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	_ "embed"
 	"encoding/json"
 	"errors"
+	"github.com/cloudogu/k8s-dogu-operator/controllers/limit"
+	mocks2 "github.com/cloudogu/k8s-dogu-operator/controllers/resource/mocks"
```
Comment (nhinze23) [inline controllers/dogu_manager_external_test.go:9]: rename package

File: controllers/limit/doguDeploymentLimitPatcher.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		return err
+	}
+
+	deployment.Spec.Template.Spec.Containers[0].Resources.Requests = resourceRequests
+	deployment.Spec.Template.Spec.Containers[0].Resources.Limits = resourceLimits
+
+	return nil
+}
+
+func (d *doguDeploymentLimitPatcher) patchStorageEphemeralLimits(limits DoguLimits, resourceRequests map[v1.ResourceName]containerresource.Quantity, resourceLimits map[v1.ResourceName]containerresource.Quantity) error {
```
Comment (nhinze23) [inline controllers/limit/doguDeploymentLimitPatcher.go:88]: use ResourceList Type

File: controllers/limit/doguDeploymentLimitPatcher_test.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+	t.Run("successfully create limits with some of the keys", func(t *testing.T) {
+		// given
+		regMock := &mocks.Registry{}
+		testDoguRegistry := &mocks.ConfigurationContext{}
+		regMock.On("DoguConfig", "testDogu").Return(testDoguRegistry)
+
+		testDoguRegistry.On("Get", cpuLimitKey).Return("100m", nil)
+		testDoguRegistry.On("Get", memoryLimitKey).Return("1Gi", nil)
+		testDoguRegistry.On("Get", ephemeralStorageLimitKey).Return("4Gi", nil)
```
Comment (nhinze23) [inline controllers/limit/doguDeploymentLimitPatcher_test.go:103]: delete one key

File: controllers/limit/doguDeploymentLimitPatcher_test.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+			}}},
+		}
+
+		regMock := &mocks.Registry{}
+		patcher := doguDeploymentLimitPatcher{}
+		patcher.registry = regMock
+
+		doguLimits := DoguLimits{
+			CpuLimit:    "100m",
+			MemoryLimit: "2Gi",
```
Comment (nhinze23) [inline controllers/limit/doguDeploymentLimitPatcher_test.go:247]: delete one limit

File: controllers/limit/hardwareLimitUpdater.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	coreosclient "github.com/coreos/etcd/client"
+	"github.com/hashicorp/go-multierror"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	triggerSyncEtcdKeyFullPath = "/config/_global/trigger-container-limit-sync"
```
Comment (nhinze23) [inline controllers/limit/hardwareLimitUpdater.go:18]: lets use underscores here too

File: controllers/limit/hardwareLimitUpdater.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	// RetrieveMemoryLimits reads all container keys from the dogu configuration and creates a DoguLimits object.
+	RetrieveMemoryLimits(doguResource *k8sv1.Dogu) (DoguLimits, error)
+	// PatchDeployment patches the given deployment with the resource limits provided.
+	PatchDeployment(deployment *v1.Deployment, limits DoguLimits) error
+}
+
+// DoguLimits contains all data necessary to limit the physical resources for a dogu.
+type DoguLimits struct {
+	// Sets the cpu requests and limit values for the dogu deployment to the contained value. For more information about resource management in Kubernetes see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/.
+	CpuLimit string
```
Comment (nhinze23) [inline controllers/limit/hardwareLimitUpdater.go:39]: unexport members

File: controllers/limit/hardwareLimitUpdater.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	// Sets the memory requests and limit values for the dogu deployment to the contained value. For more information about resource management in Kubernetes see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/.
+	MemoryLimit string
+	// Sets the ephemeral storage requests and limit values for the dogu deployment to the contained value. For more information about resource management in Kubernetes see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/.
+	EphemeralStorageLimit string
+}
+
+// NewHardwareLimitUpdater creates a new runnable responsible to detect changes in the container configuration of dogus.
+func NewHardwareLimitUpdater(client client.Client, namespace string) (*hardwareLimitUpdater, error) {
+	endpoint := fmt.Sprintf("http://etcd.%s.svc.cluster.local:4001", namespace)
```
Comment (nhinze23) [inline controllers/limit/hardwareLimitUpdater.go:48]: we could use the registry from main here

File: controllers/limit/hardwareLimitUpdater.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		}
+	}
+
+	return result
+}
+
+func (hlu *hardwareLimitUpdater) getInstalledDogus(ctx context.Context) (*k8sv1.DoguList, error) {
+	doguList := &k8sv1.DoguList{}
+
+	err := hlu.client.List(ctx, doguList, client.InNamespace(hlu.namespace))
```
Comment (nhinze23) [inline controllers/limit/hardwareLimitUpdater.go:135]: nice!

File: controllers/limit/doguDeploymentLimitPatcher.go (hunk 2 above)
Comment (Subaro) [reply controllers/limit/doguDeploymentLimitPatcher.go:88]: Done

File: controllers/limit/doguDeploymentLimitPatcher_test.go (hunk 3 above)
Comment (Subaro) [reply controllers/limit/doguDeploymentLimitPatcher_test.go:103]: Done

File: controllers/limit/doguDeploymentLimitPatcher_test.go (hunk 4 above)
Comment (Subaro) [reply controllers/limit/doguDeploymentLimitPatcher_test.go:247]: Done

File: controllers/limit/hardwareLimitUpdater.go (hunk 6 above)
Comment (Subaro) [reply controllers/limit/hardwareLimitUpdater.go:39]: Done

File: controllers/limit/hardwareLimitUpdater.go (hunk 7 above)
Comment (Subaro) [reply controllers/limit/hardwareLimitUpdater.go:48]: We could but there are no real benefits, right?

File: controllers/limit/hardwareLimitUpdater.go (hunk 8 above)
Comment (Subaro) [reply controllers/limit/hardwareLimitUpdater.go:135]: :sunglasses:

File: controllers/limit/hardwareLimitUpdater_test.go [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+		}
+
+		ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second*2)
+
+		// when
+		_ = hardwareUpdater.Start(ctx)
+		cancelFunc()
+
+		// then
+		//require.Error(t, err, assert.AnError)
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/limit/hardwareLimitUpdater_test.go:220]: Done

File: docs/operations/configuring_limits_for_a_dogu_en.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+This document describes how to configure and apply physical limits for a dogu.
+
+## Prerequisites
+
+* An operational Cloudogu MultiNode EcoSystem.
+
+## Physical Limits
+
+Physical limits can be applied to each dogu and restrict the pod of the dogu to specified limits.
+limits.
```
Comment (nhinze23) [inline docs/operations/configuring_limits_for_a_dogu_en.md:12]: delete line

File: docs/operations/configuring_limits_for_a_dogu_en.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+- Description: Sets the ephemeral storage limit for each started pod of the dogus.
+- Format:
+  See https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-units-in-kubernetes
+
+## Apply configured limits
+
+In order for the limits to be applied, the global etcd key: `config/_global/trigger_container_limit_sync`
+must be created/modified/deleted. Any change to the key will result in the start of an automatic update process
+for all dogus. In this update process the limits are applied to all dogus and the dogus are restarted if new limits are set.
+are restarted. Unchanged dogus are not restarted. In general, the update process can be started with the
```
Comment (nhinze23) [inline docs/operations/configuring_limits_for_a_dogu_en.md:58]: Delete are restarted

File: docs/operations/configuring_limits_for_a_dogu_en.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+   Kubernetes: [Kubernetes-CPU](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)
+1. **Memory Limit**: More information is available on the official page of
+   Kubernetes: [Kubernetes-Memory](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
+1. **Ephemeral-Limit**: More information is available on the official page of
+   Kubernetes: [Kubernetes-Ephemeral-Storage](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage)
+
+## Configuring Limits
+
+Limits are generally configured in etcd. **Note:** Setting a limit does not automatically cause a restart of the
+restart of the Dogus. Limits must be applied explicitly. This is described in the next section.
```
Comment (nhinze23) [inline docs/operations/configuring_limits_for_a_dogu_en.md:25]: Delete restart of the

**Pull request merged**